src/azul/portal_service.py

```diff
@@ -172,11 +172,12 @@ def _crud(self, operation: Callable[[JSONs], Optional[JSONs]]) -> Optional[JSONs
                     except NoSuchObjectVersion:
                         # Wait for latest version to appear in S3.
                         continue
-                db = operation(db)
-                if db is None:
-                    # Operation is read-only; we're done.
+                new_db = operation(db)
+                if db == new_db:
+                    # Operation didn't modify DB; we're done.
                     break
                 else:
+                    db = new_db or db
```

Comment:
Once [these changes](https://github.com/DataBiosphere/azul/pull/3461#issuecomment-930667296) are made, I think can be simplified to 
```suggestion
                    db = new_db
```